packages/debug/src/node/stream-debug-adapter.ts

```diff
@@ -16,10 +16,13 @@
 
 import { DisposableCollection } from '@theia/core/lib/common/disposable';
 import { Emitter, Event } from '@theia/core/lib/common/event';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { ChildProcess } from 'child_process';
 import * as stream from 'stream';
-import { CommunicationProvider } from './debug-model';
+import * as net from 'net';
+import { DebugAdapter } from './debug-model';
 
-export class StreamCommunicationProvider extends DisposableCollection implements CommunicationProvider {
+abstract class StreamDebugAdapter extends DisposableCollection {
```

Comment:
It would be nice to see this adapter pulled into a common area with the node specific implementations below existing in the `node` folder. This would help abstract this from any browser implementations ;)

Reply:
I agree this implementation might be interesting to reuse for browser-side debug adapters (if they exist). However, I see that as a different PR, not part of "cleaning up the lifecycle".

Reply:
We have a browser debug adapter for Keil Studio today, currently reworking it to test your PR which is where these comments are coming from.

Reply:
> I see that as a different PR, not part of "cleaning up the lifecycle".
Ack